File: AtCoder/EDPC/probF.cpp
#include <bits/stdc++.h>
typedef long long ll;
using namespace std;

int main() {
    string S, T; cin >> S >> T;
    int ls = S.length(), lt = T.length();
    int dp[ls+1][lt+1];
    for (int i=0; i<=ls; i++){
        dp[i][0] = 0;
    }
    for (int i=0; i<=lt; i++){
        dp[0][i] = 0;
    }
    for (int i=1; i<=ls; i++){
        for (int j=1; j<=lt; j++){
            if (S[i-1] == T[j-1]){
                dp[i][j] = dp[i-1][j-1] + 1;
            }
            else{
                dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
            }
        }
    }

    int i = ls-1, j = lt-1;
    bool update;
    string ans = "";
    while (i >= 0 && j >= 0){
        if (S[i] == T[j]) {
            ans = S[i] + ans;
            i -= 1;
            j -= 1;
        }
        else {
            if (dp[i][j+1] > dp[i+1][j]){
                i -= 1;
            }
            else {
                j -= 1;
            }
        }
    }
    cout << ans << endl;
}

File: AtCoder/ABC179/probF.py
INF = 10**18

class BITmin():
    def __init__(self, max):
        self.max = max
        self.data = [self.max-1]*(self.max+1)
    
    # 上に登ってく
    def query_min(self, i):
        s = INF
        while i <= self.max:
            if s > self.data[i]:
                s = self.data[i]
            i += i & -i
        return s

    # iまで全ての要素をxで更新
    def update(self, i, x):
        while i > 0:
            if x < self.data[i]:
                self.data[i] = x
            # i += i & -i
            i -= i & -i

import sys
input = sys.stdin.buffer.readline

N, Q = map(int, input().split())
Query = [list(map(int, input().split())) for _ in range(Q)]

BitX = BITmin(N)
BitY = BITmin(N)

ans = (N-2)*(N-2)
for a, b in Query:
    b -= 1
    if a == 1:
        loc = BitX.query_min(b)
        ans -= loc-1
        BitY.update(loc, b)
    else:
        loc = BitY.query_min(b)
        ans -= loc-1
        BitX.update(loc, b)
    
print(ans)